File: app/src/main/java/com/md/modesetters/DeckChooseModeSetter.java
package com.md.modesetters;

import android.app.Activity;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.AdapterView.OnItemLongClickListener;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import com.md.CategorySingleton;
import com.md.DbNoteEditor;
import com.md.ModeHandler;
import com.md.R;
import com.md.RevisionQueue;
import com.md.modesetters.deckchoose.DeckDeleter;
import com.md.modesetters.deckchoose.DeckNameUpdater;
import com.md.modesetters.deckchoose.InsertNewHandler;
import com.md.provider.Deck;
import com.md.utils.ToastSingleton;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Vector;

public class DeckChooseModeSetter extends ModeSetter {

	private static DeckChooseModeSetter instance = null;
	private int mTotalNotes;

	protected DeckChooseModeSetter() {
		deckInfoMap = new LinkedHashMap<Integer, DeckInfo>();
	}

	public static DeckChooseModeSetter getInstance() {
		if (instance == null) {
			instance = new DeckChooseModeSetter();
		}
		return instance;
	}

	private Activity memoryDroid;

	public void setUp(Activity memoryDroid,
			ModeHandler modeHand) {
		parentSetup(memoryDroid, modeHand);
		this.memoryDroid = memoryDroid;
	}

	public void setupModeImpl(final Activity context) {
		commonSetup(context, R.layout.deckchoosetemp);
		setupCreateMode();
	}

	private ListView mListView;
	private Vector<Deck> mAdapterDecks = new Vector<Deck>();
	private HashMap<Integer, DeckInfo> deckInfoMap;
	private boolean loadComplete;
	private ProgressBar progressBar;

	public void setupCreateMode() {

		loadComplete = false;

		Button insertButton = (Button) memoryDroid.findViewById(R.id.addNew);

		insertButton.setOnClickListener(new InsertNewHandler(memoryDroid,
				instance));

		mAdapterDecks.clear();

		Vector<Deck> queryDeck = DbNoteEditor.getInstance().queryDeck();

		for (Deck deck : queryDeck) {
			addDeck(deck.getName(), deck.getId());
		}

		setListView((ListView) memoryDroid.findViewById(R.id.ListView01));

		progressBar = (ProgressBar) memoryDroid.findViewById(R.id.progressBar);

		progressBar.setMax(mAdapterDecks.size());

		progressBar.setProgress(0);

		progressBar.setVisibility(View.VISIBLE);

		ArrayAdapter<Deck> arrayAdapter = new ArrayAdapter<Deck>(memoryDroid,
				android.R.layout.simple_list_item_1, mAdapterDecks);

		// By using setAdpater method in listview we an add string array in
		// list.
		getListView().setAdapter(arrayAdapter);

		OnItemClickListener onItemClickListener = new OnItemClickListenerImplementation(
				memoryDroid);
		getListView().setOnItemClickListener(onItemClickListener);
		OnItemLongClickListener onItemLongClickListener = new OnItemLongClickListener() {

			@Override
			public boolean onItemLongClick(AdapterView<?> av, final View v,
										   int index, long arg) {
				if (!loadComplete) {
					return true;
				}
				final DeckInfo deckInfo = deckInfoMap.get((int) arg);
				AlertDialog.Builder alert = new AlertDialog.Builder(memoryDroid);
				alert.setTitle("Choose action or press off screen");
				alert.setPositiveButton("Rename", new DialogInterface.OnClickListener() {
					public void onClick(DialogInterface dialog, int whichButton) {
						new DeckNameUpdater(memoryDroid, deckInfo, instance).onClick(v);
					}
				});

				alert.setNegativeButton("Delete", new DialogInterface.OnClickListener() {
					public void onClick(DialogInterface dialog, int whichButton) {
						new DeckDeleter(memoryDroid, deckInfo, instance).onClick(v);
					}
				});
				alert.show();
				return true;
			}
		};

		getListView().setOnItemLongClickListener(onItemLongClickListener);

		DeckPopulator deckPopulator = new DeckPopulator(this);

		deckPopulator.execute(this);
	}

	private final class OnItemClickListenerImplementation implements
			OnItemClickListener {

		private OnItemClickListenerImplementation(Activity memoryDroid) {

		}

		@Override
		public void onItemClick(AdapterView<?> av, View v, int index, long arg) {
			if (loadComplete) {
				DeckInfo deckInfo = deckInfoMap.get(Integer.valueOf((int) arg));
				loadDeck(deckInfo);
			}
		}
	}

	public class DeckPopulator extends
			AsyncTask<DeckChooseModeSetter, DeckInfo, DeckInfo> {

		private DeckChooseModeSetter dcChooseModeSetter;

		public DeckPopulator(DeckChooseModeSetter dcChooseModeSetter) {
			this.dcChooseModeSetter = dcChooseModeSetter;

		}

		protected void onPostExecute(DeckInfo result) {
			if (modeHand.whoseOnTop() == dcChooseModeSetter) {
				dcChooseModeSetter.onComplete();
			}
		}

		public void publishProgessVisible(DeckInfo state) {
			// Only do stuff if I'm still in control.
			if (modeHand.whoseOnTop() == dcChooseModeSetter) {
				publishProgress(state);
			}
		}

		protected void onProgressUpdate(DeckInfo... state) {
			dcChooseModeSetter.setState(state[0]);
		}

		@Override
		protected DeckInfo doInBackground(DeckChooseModeSetter... params) {

			mTotalNotes = 0;
			int childCount = mAdapterDecks.size();

			for (int idx = 0; idx < childCount; idx++) {
				Deck elementAt = mAdapterDecks.elementAt(idx);

				// Stop if we aren't loaded anymore.
				if (modeHand.whoseOnTop() != dcChooseModeSetter) {
					break;
				}

				RevisionQueue revisionQueue = new RevisionQueue();
				revisionQueue.populate(DbNoteEditor.getInstance(),
						elementAt.getId());

				// Stop if we aren't loaded anymore. We want this before
				// and after the query
				if (modeHand.whoseOnTop() != dcChooseModeSetter) {
					break;
				}

				// TODO why is the counting when we have the size.
				int deckCount = DbNoteEditor.getInstance().getDeckCount(
						elementAt.getId());

				DeckInfo deckInfo = new DeckInfo(elementAt, revisionQueue,
						deckCount);

				publishProgessVisible(deckInfo);
			}
			return null;
		}
	}

	private void addDeck(String name, int id) {
		Deck deck = new Deck(id, name);
		mAdapterDecks.add(deck);
	}

	public void onComplete() {
		TextView loadingOrSelect = (TextView) memoryDroid
				.findViewById(R.id.loadingOrSelect);

		loadingOrSelect.setText("Press or Long Press a Deck");

		loadComplete = true;

		ToastSingleton.getInstance().msgCommon(mTotalNotes + " notes!", 0);
	}

	public void setState(DeckInfo state) {
		progressBar.setProgress(progressBar.getProgress() + 1);

		for (int idx = 0; idx < mAdapterDecks.size(); idx++) {
			Deck elementAt = mAdapterDecks.elementAt(idx);
			if (elementAt.getId() == state.getCategory()) {
				elementAt.setSize(state.getDeckCount());
				elementAt.setTodayReview(state.getRevisionQueue().getSize());
				deckInfoMap.put(Integer.valueOf(idx), state);
				// TODO(jrochest) Why is the necessary. It only needs it when returning from
				// a different mode, but it is not needed initially.
				if (mListView.getChildCount() - 1 >= idx) {
					((TextView)mListView.getChildAt(idx)).setText(state.getDeck().toString());
				}
			}
		}
	}

	public void setListView(ListView listView) {
		this.mListView = listView;
	}

	public ListView getListView() {
		return mListView;
	}

	public void loadDeck(DeckInfo deckInfo) {

		if (deckInfo != null) {
			CategorySingleton.getInstance().setDeckInfo(deckInfo);

			RevisionQueue.getInstance().makeThisLookLikeThat(
					deckInfo.getRevisionQueue());
		}

		LearningModeSetter.getInstance().setupMode(memoryDroid);
	}

    public DeckInfo getDefaultDeck() {
        if (!deckInfoMap.isEmpty()) {
            // Return the first and the default.
            return deckInfoMap.entrySet().iterator().next().getValue();
        }
        return null;
    }

}

File: app/src/main/java/com/md/modesetters/DeckChooseModeSetter.kt
package com.md.modesetters

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import com.md.*
import com.md.composeModes.CurrentNotePartManager
import com.md.composeModes.DeckMode
import com.md.composeModes.DeckModeStateModel
import com.md.composeModes.Mode
import com.md.composeModes.TopModeViewModel
import com.md.modesetters.DeckItemPopulator.populate
import com.md.modesetters.deckchoose.DeckNameUpdater
import com.md.modesetters.deckchoose.InsertNewHandler
import com.md.provider.Deck
import com.md.utils.ToastSingleton
import dagger.hilt.android.qualifiers.ActivityContext
import dagger.hilt.android.scopes.ActivityScoped
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*
import javax.inject.Inject


@ActivityScoped
class DeckLoadManager @Inject constructor(
    private val topModeViewModel: TopModeViewModel,
    private val deckModeStateModel: DeckModeStateModel,
    @ActivityContext val context: Context,
    val currentNotePartManager: CurrentNotePartManager,
    private val focusedQueueStateModel: FocusedQueueStateModel,) {
    val decks = MutableStateFlow<List<DeckInfo>?>(null)
    var deckRefreshNeeded = true

    val activity: SpacedRepeaterActivity by lazy {
        context as SpacedRepeaterActivity
    }

    init {
        activity.lifecycleScope.launch {
            refreshDeckListAndFocusFirstActiveNonemptyQueue(true)
        }

        activity.lifecycleScope.launch {
            decks.collect { decks ->
                if (decks.isNullOrEmpty()) {
                    return@collect
                }
                chooseDeck(decks)
            }
        }
    }

    fun chooseDeck() {
        val decks = decks.value ?: return
        var foundNonEmptyDeck = false
        for (deckInfo in decks) {
            if (deckInfo.isActive && !deckInfo.revisionQueue.isEmpty()) {
                setDeck(deckInfo)
                foundNonEmptyDeck = true
                break
            }
        }
        if (foundNonEmptyDeck) {
            return
        } // else use an empty deck.
        for (deckInfo in decks) {
            if (deckInfo.isActive) {
                setDeck(deckInfo)
                break
            }
        }
    }

    fun chooseDeck(decks: List<DeckInfo>) {
        for (deckInfo in decks) {
            if (deckInfo.isActive && !deckInfo.revisionQueue.isEmpty()) {
                setDeck(deckInfo)
                break
            }
        }
    }

    private fun setDeck(deckInfo: DeckInfo) {
        CategorySingleton.getInstance().setDeckInfo(deckInfo)
        focusedQueueStateModel.deck.value = deckInfo
        val note = deckInfo.revisionQueue.peekQueue()
        currentNotePartManager.changeCurrentNotePart(note, partIsAnswer = false)
    }

    suspend fun refreshDeckListAndFocusFirstActiveNonemptyQueue(deckRefreshNeeded: Boolean? = null) {
        deckRefreshNeeded?.let { this.deckRefreshNeeded = it }
        withContext(Dispatchers.Main) {
            // Only run this refresh once.
            if (this@DeckLoadManager.deckRefreshNeeded) {
                this@DeckLoadManager.deckRefreshNeeded = false
                val resultingDeckList = withContext(Dispatchers.IO) {
                    val resultingDeckList = mutableListOf<DeckInfo>()
                    val deckList = mutableListOf<Deck>()
                    val queryDeck = DbNoteEditor.instance!!.queryDeck()
                    for (deck in queryDeck) {
                        deckList.add(Deck(deck.id, deck.name))
                    }

                    for (deck in deckList) {
                        val revisionQueue = RevisionQueue()
                        revisionQueue.populate(DbNoteEditor.instance!!, deck.id)
                        val deckCount = DbNoteEditor.instance!!.getDeckCount(deck.id)
                        val deckInfo = DeckInfo(deck, revisionQueue, deckCount)
                        resultingDeckList.add(deckInfo)
                    }

                    withContext(Dispatchers.Main) {
                        decks.value = resultingDeckList
                    }
                    resultingDeckList
                }

                val decks = resultingDeckList
                if (decks.isEmpty()) {
                    withContext(Dispatchers.Main) {
                        // Go to Deck Chooser to add a deck.
                        topModeViewModel.modeModel.value = Mode.DeckChooser
                        deckModeStateModel.modeModel.value = DeckMode.AddingDeck
                    }
                    return@withContext
                } else {
                    withContext(Dispatchers.Main) {
                        chooseDeck()
                    }
                }
            }
        }
    }
}
//TODOJNOW hook up the proceed button.

@SuppressLint("StaticFieldLeak")
object DeckChooseModeSetter : ModeSetter() {
    private var mTotalNotes = 0
    @SuppressLint("StaticFieldLeak")
    private var memoryDroid: SpacedRepeaterActivity? = null
    fun setUp(memoryDroid: SpacedRepeaterActivity?,
              modeHand: ModeHandler?) {
        parentSetup(memoryDroid, modeHand)
        this.memoryDroid = memoryDroid
    }

    override fun onSwitchToMode(context: Activity) {
        commonSetup(context, R.layout.deckchoosetemp)
        setupCreateMode()
    }

    var listView: ListView? = null
    private val deckList = Vector<Deck>()
    private var deckIdToInfo: HashMap<Int, DeckInfo> = LinkedHashMap()
    private var loadComplete = false
    private var progressBar: ProgressBar? = null
    fun setupCreateMode() {
        val activity = checkNotNull(mActivity)

        loadComplete = false
        val insertButton = memoryDroid!!.findViewById<View>(R.id.addNew) as Button
        insertButton.setOnClickListener(InsertNewHandler(memoryDroid, this))
        deckList.clear()

        listView = memoryDroid!!.findViewById<View>(R.id.ListView01) as ListView
        progressBar = memoryDroid!!.findViewById<View>(R.id.progressBar) as ProgressBar
        progressBar!!.max = deckList.size
        progressBar!!.progress = 0
        progressBar!!.visibility = View.VISIBLE

        // By using setAdpater method in listview we an add string array in
        // list.
        listView!!.adapter = DeckAdapter(deckList, mActivity!!)

        //val deckPopulator = DeckPopulator(this)
        //deckPopulator.execute(this)
        val modeChooser = this
    }

    private class DeckAdapter(
        private val decks: List<Deck>,
        private val context: Activity,
    ) : BaseAdapter() {
        override fun getCount(): Int {
            return decks.size
        }

        override fun getItem(position: Int): Any {
            return decks[position]
        }

        override fun getItemId(position: Int): Long {
            return decks[position].id.toLong()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view = convertView ?: context.layoutInflater.inflate(R.layout.deck_item_for_list, parent, false)

            val deck = decks[position]
            val deckInfo = deckIdToInfo[deck.id]
            val textView = view.findViewById<TextView>(R.id.deck_name)
            if (deckInfo != null) {
                populate(view, deckInfo)
            } else {
                textView.text = deck.name
            }
            textView.setOnLongClickListener { v ->
                val info = deckIdToInfo[deck.id] ?: return@setOnLongClickListener true
                val alert = AlertDialog.Builder(memoryDroid!!)
                alert.setTitle("Choose action or press off screen")
                alert.setPositiveButton("Rename") { _, _ -> DeckNameUpdater(memoryDroid, info, getInstance()).onClick(v) }
                //alert.setNegativeButton("Delete") { _, _ -> DeckDeleter(memoryDroid, info, getInstance()).onClick(v) }
                alert.show()
                true
            }
            view.findViewById<View>(R.id.new_note_button).setOnClickListener {
                val info = deckIdToInfo[deck.id] ?: return@setOnClickListener
                loadDeck(info)
                CreateModeSetter.switchMode(memoryDroid!!)
            }

                view.findViewById<Button>(R.id.learn_button).setOnClickListener {
                val info = deckIdToInfo[deck.id] ?: return@setOnClickListener
                loadDeck(info)
                    memoryDroid?.switchToLearningMode()
            }

            return view
        }
    }

    fun onComplete() {
        val activity = checkNotNull(memoryDroid)
        val deckLoadManager = activity.deckLoadManager()

        activity.lifecycleScope.launch {


        }

        val loadingOrSelect = memoryDroid?.findViewById<View>(R.id.loadingOrSelect) as TextView
        loadingOrSelect.text = "Press or Long Press a Deck"
        loadComplete = true
        ToastSingleton.getInstance().msgCommon("$mTotalNotes notes!")
    }

    fun applyDeckInfoToExistingUiElement(state: DeckInfo) {
        progressBar!!.progress = progressBar!!.progress + 1
        for (idx in deckList.indices) {
            val elementAt = deckList.elementAt(idx)
            if (elementAt.id == state.category) {
                elementAt.setSize(state.deckCount)
                elementAt.setTodayReview(state.revisionQueue.getSize())
                deckIdToInfo[elementAt.id] = state
                // TODO(jrochest) Why is the necessary. It only needs it when returning from
                // a different mode, but it is not needed initially.
                if (listView!!.childCount - 1 >= idx) {
                    populate(listView!!.getChildAt(idx), state)
                }
            }
        }
    }

    fun loadDeck(deckInfo: DeckInfo?) {
        if (deckInfo != null) {
            CategorySingleton.getInstance().setDeckInfo(deckInfo)
        }
    }

    val nextDeckWithItems: DeckInfo?
        get() {
            if (deckIdToInfo.isEmpty()) {
                return null
            }
            for ((_, value) in deckIdToInfo) {
                if (!value.isActive) {
                    continue
                }
                if (value.revisionQueue.getSize() > 0) {
                    return value
                }
            }
            return null
        }

        fun getInstance() = this
}